execution/scan/matrix_selector.go

```diff
@@ -180,7 +180,13 @@ func (o *matrixSelector) loadSeries(ctx context.Context) error {
 		for i, s := range series {
 			lbls := s.Labels()
 			if o.funcExpr.Func.Name != "last_over_time" {
-				lbls = function.DropMetricName(lbls)
+				// This modifies the array in place. Because labels.Labels
+				// can be re-used between different Select() calls, it means that
+				// we have to copy it here.
+				// TODO(GiedriusS): could we identify somehow whether labels.Labels
+				// is reused between Select() calls?
+				lbls = function.DropMetricName(lbls.Copy())
```

Comment:
I found out that labels can be reused across Select() calls however making a copy each time is costly. Perhaps @fpetkovski you have some ideas on how to solve this better?

Reply:
Hmm so if I understand correctly, a downstream operator can modify the labels from its upstream, and thereby modify the labels of the upstream as well. I assume this is because we modify the slice in place, something that @saswatamcode implemented to reduce memory usage. 

I think the correct thing would be to make a copy after all. Otherwise we can end up with this sort of subtle bugs that are hard to track. When you say it's costly to copy the labels, did you see anything in a benchmark that stood out?

Reply:
Hmm, I thought that labels were essentially "handed-off" to downstream operators but that doesn't seem to be the case.

Reply:
I feel we should maintain the semantics for every operators series function to prevent this kind of bug in the future.

Should we always do a labels copy if labels are modified?